Unable to parse `error code` if `source` file has uncomplete error msg

# Reproducer

I tried this code from https://github.com/rust-lang/rust/blob/master/tests/ui/associated-types/issue-20005.rs

#### `test.rs`

```rust
trait From<Src> {
    type Result;

    fn from(src: Src) -> Self::Result;
}

trait To {
    fn to<Dst>(
        self //~ ERROR the size for values of type
    ) -> <Dst as From<Self>>::Result where Dst: From<Self> { //~ ERROR the size for values of type
        From::from(self) //~ ERROR the size for values of type
    }
}

fn main() {}
```

<details><summary>test.stderr</summary>
<p>

```rust
error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/issue-20005.rs:10:49
   |
LL |     ) -> <Dst as From<Self>>::Result where Dst: From<Self> {
   |                                                 ^^^^^^^^^^ doesn't have a size known at compile-time
   |
note: required by an implicit `Sized` bound in `From`
  --> $DIR/issue-20005.rs:1:12
   |
LL | trait From<Src> {
   |            ^^^ required by the implicit `Sized` requirement on this type parameter in `From`
help: consider further restricting `Self`
   |
LL |     ) -> <Dst as From<Self>>::Result where Dst: From<Self>, Self: Sized {
   |                                                           +++++++++++++
help: consider relaxing the implicit `Sized` restriction
   |
LL | trait From<Src: ?Sized> {
   |               ++++++++

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/issue-20005.rs:9:9
   |
LL |         self
   |         ^^^^ doesn't have a size known at compile-time
   |
   = help: unsized fn params are gated as an unstable feature
help: consider further restricting `Self`
   |
LL |     ) -> <Dst as From<Self>>::Result where Dst: From<Self>, Self: Sized {
   |                                                           +++++++++++++
help: function arguments must have a statically known size, borrowed types always have a known size
   |
LL |         &self
   |         +

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/issue-20005.rs:11:9
   |
LL |         From::from(self)
   |         ^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
note: required by an implicit `Sized` bound in `From`
  --> $DIR/issue-20005.rs:1:12
   |
LL | trait From<Src> {
   |            ^^^ required by the implicit `Sized` requirement on this type parameter in `From`
help: consider further restricting `Self`
   |
LL |     ) -> <Dst as From<Self>>::Result where Dst: From<Self>, Self: Sized {
   |                                                           +++++++++++++
help: consider relaxing the implicit `Sized` restriction
   |
LL | trait From<Src: ?Sized> {
   |               ++++++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
```


</p>
</details> 

# Expected behavior:

```rust
trait From<Src> {
    type Result;

    fn from(src: Src) -> Self::Result;
}

trait To {
    fn to<Dst>(
        self // { dg-error ".E0277." "" { target *-*-* } }
    ) -> <Dst as From<Self>>::Result where Dst: From<Self> { // { dg-error ".E0277." "" { target *-*-* } }
        From::from(self) // { dg-error ".E0277." "" { target *-*-* } }
    }
}

fn main() {}

```

# Actual behavior:

- No error codes

```rust
trait From<Src> {
    type Result;

    fn from(src: Src) -> Self::Result;
}

trait To {
    fn to<Dst>(
        self // { dg-error "" "" { target *-*-* } }
    ) -> <Dst as From<Self>>::Result where Dst: From<Self> { // { dg-error "" "" { target *-*-* } }
        From::from(self) // { dg-error "" "" { target *-*-* } }
    }
}

fn main() {}

```
### Tool version:

[f91da539e62db42c51f74022becdbff901bff1a2](https://github.com/Rust-GCC/rusttest-to-dg/pull/6/commits/f91da539e62db42c51f74022becdbff901bff1a2)

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,4 +8,5 @@ edition = "2021"
 [dependencies]
 anyhow = "1.0.86"
 clap = { version = "4.5.8", features = ["derive"] }
+once_cell = "1.19.0"
 regex = "1.10.5"

--- a/src/cli.rs
+++ b/src/cli.rs
@@ -1,15 +1,19 @@
+//! This module contains the command line interface for the tool
+
 use {
     anyhow::{Context, Result},
     clap::Parser,
     std::{fs, path},
 };
 
+/// Command line arguments for the tool
 #[derive(Parser, Debug)]
 #[command(
     name = "rust test to DejaGnu",
     long_about = "A tool to convert rust tests into DejaGnu tests format"
 )]
 pub struct Arguments {
+    /// The rust source file to convert into `DejaGnu` format
     #[arg(
         short = 'f',
         long = "file",
@@ -18,6 +22,7 @@ pub struct Arguments {
     )]
     pub source_file: path::PathBuf,
 
+    /// optional `stderr` file
     #[arg(
         short = 'e',
         long = "stderr",
@@ -33,6 +38,7 @@ pub fn parse_arguments_and_read_file(args: &Arguments) -> Result<(String, Option
     let source_code = fs::read_to_string(&args.source_file)
         .with_context(|| format!("could not read sourcefile `{}`", args.source_file.display()))?;
 
+    // Read the stderr file if it exists
     let err_file =
         match &args.stderr_file {
             Some(stderr_file) => Some(fs::read_to_string(stderr_file).with_context(|| {
@@ -66,6 +72,8 @@ mod tests {
         assert_eq!(args.stderr_file, Some(path::PathBuf::from("test.stderr")));
     }
 
+    /// clap reports most development errors as `debug_assert!`s
+    /// See this for more details, [here](https://docs.rs/clap/4.5.15/clap/_derive/_tutorial/chapter_4/index.html)
     #[test]
     fn debug_args() {
         use clap::CommandFactory;

--- a/src/cli.rs
+++ b/src/cli.rs
@@ -1,7 +1,8 @@
-use std::{fs, path};
-
-use anyhow::{Context, Result};
-use clap::Parser;
+use {
+    anyhow::{Context, Result},
+    clap::Parser,
+    std::{fs, path},
+};
 
 #[derive(Parser, Debug)]
 #[command(

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,3 +1,5 @@
+//! This module contains the logic for parsing rustc error messages.
+
 use {
     self::WhichLine::*,
     std::{fmt, str::FromStr},
@@ -11,7 +13,9 @@ macro_rules! regex {
         RE.get_or_init(|| regex::Regex::new($re).unwrap())
     }};
 }
-// https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-levels
+
+/// Represents the different kinds of Rustc compiler messages.
+/// See [rustc dev guide](https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-levels)
 #[derive(Copy, Clone, Debug, PartialEq)]
 pub enum RustcErrorKind {
     Help,
@@ -55,6 +59,7 @@ impl fmt::Display for RustcErrorKind {
     }
 }
 
+/// To store information from rustc source file
 #[derive(Debug)]
 pub struct Error {
     pub line_num: usize,
@@ -67,11 +72,14 @@ pub struct Error {
     /// What kind of message we expect (e.g., warning, error, suggestion).
     /// `None` if not specified or unknown message kind.
     pub kind: Option<RustcErrorKind>,
+    ///Note: if we are loading this from rustc source file, this might be incomplete
     pub msg: String,
     pub error_code: Option<String>,
 }
 
 impl fmt::Display for Error {
+    /// Formats the `Error` for display according to `DejaGnu` format
+    /// See `DejaGnu` documentation [here](https://gcc.gnu.org/onlinedocs/gccint/testsuites/directives-used-within-dejagnu-tests/syntax-and-descriptions-of-test-directives.html)
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use RustcErrorKind::*;
 
@@ -105,6 +113,9 @@ impl fmt::Display for Error {
     }
 }
 
+/// Represents the line in the rustc source code where an error occurred.
+/// Luckily, rust compile test only stores error messages on and after the line where the error occurred.
+/// But `DejaGnu` can process error messages on the previous line, the current line, or the next line.
 #[derive(PartialEq, Debug)]
 enum WhichLine {
     ThisLine,
@@ -112,8 +123,10 @@ enum WhichLine {
     AdjustBackward(usize),
 }
 
+/// The main function for loading errors from source file and from optional stderr file.
 pub fn load_error(text_file: &str, stderr_file: Option<&str>) -> Vec<Error> {
     let mut last_unfollow_error = None;
+    // For storing the errors
     let mut errors = Vec::new();
 
     for (line_num, line) in text_file.lines().enumerate() {
@@ -126,12 +139,17 @@ pub fn load_error(text_file: &str, stderr_file: Option<&str>) -> Vec<Error> {
         }
     }
 
+    // If stderr file is not provided, return the errors
     if stderr_file.is_none() {
         return errors;
     }
     // TODO: improve this code incrementally
+    // parsing error related information from `.stderr` file
     let error_code_stderr = parse_error_code(stderr_file.expect("stderr file is not found"));
 
+    // TODO: We need to load error messages from `.stderr` instead of source file become sometimes source file contains incomplete error messages
+    // finding the error code w.r.t line number and error message
+    // TODO: sometimes, the error message might not be same but this doesn't matter as we are not comparing the row number for the message
     for error in errors.iter_mut() {
         for error_code in error_code_stderr.iter() {
             if error.line_num == error_code.line_number
@@ -141,9 +159,11 @@ pub fn load_error(text_file: &str, stderr_file: Option<&str>) -> Vec<Error> {
             }
         }
     }
+    // return error detail with error codes
     errors
 }
 
+/// To represent information from `stderr` file
 #[derive(Debug)]
 struct StderrResult {
     error_code: String,
@@ -155,6 +175,7 @@ fn is_error_code(s: &str) -> bool {
     regex!(r"^E\d{4}$").is_match(s)
 }
 
+/// Parses error codes from the `stderr` file
 fn parse_error_code(stderr_content: &str) -> Vec<StderrResult> {
     // Modified regex pattern with named capture groups
     let error_pattern = regex!(
@@ -191,6 +212,7 @@ fn parse_error_code(stderr_content: &str) -> Vec<StderrResult> {
     results
 }
 
+/// Parses error details from a source line.
 fn parse_expected(
     last_nonfollow_error: Option<usize>,
     line_num: usize,
@@ -228,6 +250,7 @@ fn parse_expected(
 
     let msg = msg.trim().to_owned();
 
+    // If we find `//~|` or `//~^`, we need to adjust the line number.
     let mut relative_line_num = line_num as i32;
     let (which, line_num) = if follow {
         assert_eq!(adjusts, 0, "use either //~| or //~^, not both.");

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -135,7 +135,7 @@ pub fn load_error(text_file: &str, stderr_file: Option<&str>) -> Vec<Error> {
     for error in errors.iter_mut() {
         for error_code in error_code_stderr.iter() {
             if error.line_num == error_code.line_number
-                && error.msg == error_code.error_message_detail
+                || error.msg == error_code.error_message_detail
             {
                 error.error_code = Some(error_code.error_code.clone());
             }

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -4,6 +4,14 @@ use regex::Regex;
 
 use self::WhichLine::*;
 
+// https://docs.rs/once_cell/1.19.0/once_cell/#lazily-compiled-regex
+#[macro_export]
+macro_rules! regex {
+    ($re:literal $(,)?) => {{
+        static RE: once_cell::sync::OnceCell<regex::Regex> = once_cell::sync::OnceCell::new();
+        RE.get_or_init(|| regex::Regex::new($re).unwrap())
+    }};
+}
 // https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-levels
 #[derive(Copy, Clone, Debug, PartialEq)]
 pub enum RustcErrorKind {

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,8 +1,7 @@
-use std::{cell::OnceCell, fmt, str::FromStr};
-
-use regex::Regex;
-
-use self::WhichLine::*;
+use {
+    self::WhichLine::*,
+    std::{fmt, str::FromStr},
+};
 
 // https://docs.rs/once_cell/1.19.0/once_cell/#lazily-compiled-regex
 #[macro_export]
@@ -153,17 +152,14 @@ struct StderrResult {
 }
 
 fn is_error_code(s: &str) -> bool {
-    let re: OnceCell<Regex> = OnceCell::new();
-    let regex = re.get_or_init(|| Regex::new(r"^E\d{4}$").unwrap());
-    regex.is_match(s)
+    regex!(r"^E\d{4}$").is_match(s)
 }
 
 fn parse_error_code(stderr_content: &str) -> Vec<StderrResult> {
     // Modified regex pattern with named capture groups
-    let re: OnceCell<Regex> = OnceCell::new();
-    let error_pattern = re.get_or_init(|| {
-        Regex::new(r"error\[(?P<error_code>E\d{4})\]: (?P<error_message_detail>.+?)\n\s+-->.+:(?P<line_number>\d+):").unwrap()
-    });
+    let error_pattern = regex!(
+        r"error\[(?P<error_code>E\d{4})\]: (?P<error_message_detail>.+?)\n\s+-->.+:(?P<line_number>\d+):"
+    );
 
     let mut results = Vec::new();
 
@@ -205,11 +201,8 @@ fn parse_expected(
     //     //~|
     //     //~^
     //     //~^^^^^
-    let re: OnceCell<Regex> = OnceCell::new();
 
-    let captures = re
-        .get_or_init(|| Regex::new(r"//(?:\[(?P<revs>[\w\-,]+)])?~(?P<adjust>\||\^*)").unwrap())
-        .captures(line)?;
+    let captures = regex!(r"//(?:\[(?P<revs>[\w\-,]+)])?~(?P<adjust>\||\^*)").captures(line)?;
 
     let (follow, adjusts) = match &captures["adjust"] {
         "|" => (true, 0),

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,3 +1,5 @@
+//! The main entry point of the program.
+
 use {
     anyhow::{Context, Result},
     clap::Parser,

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,5 +1,7 @@
-use anyhow::{Context, Result};
-use clap::Parser;
+use {
+    anyhow::{Context, Result},
+    clap::Parser,
+};
 
 mod cli;
 mod errors;

--- a/src/transform.rs
+++ b/src/transform.rs
@@ -1,9 +1,7 @@
-use std::cell::OnceCell;
-
-use anyhow::Result;
-use regex::Regex;
-
-use crate::errors;
+use {
+    crate::{errors, regex},
+    anyhow::Result,
+};
 
 /// This function takes the rust code as input
 /// and returns the code with DejaGnu directive
@@ -26,12 +24,7 @@ pub fn transform_code(code: &str, stderr_file: Option<&str>) -> Result<String> {
                 new_line = format!("{}", error);
             } else {
                 // For the error on the same line
-                let re: OnceCell<Regex> = OnceCell::new();
-
-                let captures = re
-                    .get_or_init(|| {
-                        Regex::new(r"//(?:\[(?P<revs>[\w\-,]+)])?~(?P<adjust>\||\^*)").unwrap()
-                    })
+                let captures = regex!(r"//(?:\[(?P<revs>[\w\-,]+)])?~(?P<adjust>\||\^*)")
                     .captures(line)
                     .expect("Could not find the error directive");
 

--- a/src/transform.rs
+++ b/src/transform.rs
@@ -1,36 +1,47 @@
+//! This module contains the code transformation logic.
+
 use {
     crate::{errors, regex},
     anyhow::Result,
 };
 
-/// This function takes the rust code as input
-/// and returns the code with DejaGnu directive
+/// Transform code to `DejaGnu` format
 pub fn transform_code(code: &str, stderr_file: Option<&str>) -> Result<String> {
+    // Load the rustc error messages, codes, lines and relative line numbers
     let errors = errors::load_error(code, stderr_file);
+    // For storing the transformed code
     let mut new_code = String::new();
 
     let mut line_num = 1;
+    // finding the respective line number and adding the error code
     for line in code.lines() {
         let mut new_line = line.to_string();
         // TODO: This is not the efficient way to find respective line number
         for error in errors.iter() {
+            // Checking the original line number
             if (error.line_num as i32 - error.relative_line_num) != line_num {
                 continue;
             }
             // In rustc test suites, the error directive is
-            // on the same line or the next line not on the previous line
+            // on the same line or on the next line, but not on the previous line
+            // See this: https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-annotations
             // For the error on the next line
             if error.relative_line_num != 0 {
+                // We simply add the error message, not to worry about the code
+                // The error was printed by our overloaded `Display` trait
                 new_line = format!("{}", error);
             } else {
-                // For the error on the same line
+                // For the error on the same line, we need to add error message at the end of the line
                 let captures = regex!(r"//(?:\[(?P<revs>[\w\-,]+)])?~(?P<adjust>\||\^*)")
                     .captures(line)
                     .expect("Could not find the error directive");
 
                 // Get the part of comment before the sigil (e.g. `~^` or ~|)
                 let whole_match = captures.get(0).unwrap();
+                // Get the existing source code before the error directive //~ ERROR or similar to this
                 let before_match = &line[..whole_match.start()];
+
+                // The error was printed by our overloaded `Display` trait
                 new_line = format!("{}{}", before_match, error);
             }
             break;

